uniform_isogeny_primes.py

```diff
@@ -33,36 +33,48 @@
 import argparse
 import logging
 
-from sage.all import Integer, prime_divisors
+from sage.all import Integer, prime_divisors, is_prime
 
-from sage_code.strong_uniform_bounds import unif_bd, type_one_unif_primes
+from sage_code.common_utils import is_b_smooth
+from sage_code.strong_uniform_bounds import unif_bd, type_one_unif_bound
 
 
-def do_uniform_quadratic():
+def do_uniform(d, trial_division_bound, aux_bound):
 
-    EPSILONS = [(0,4), (0,8), (4,4), (4,6), (4,12)]
+    epsilons = [
+        tuple([trace] + [0] * (d - 1)) for trace in range(0, 6 * d, 2) if trace != 2 and (trace == 0 or trace % 6 != 0)
+    ]
 
-    for eps in EPSILONS:
-        bd = Integer(unif_bd(2,eps))
-        logging.info(f"Isogeny primes for {eps} are {bd.prime_divisors()}")
+    for eps in epsilons:
+        do_uniform_eps(d, eps, trial_division_bound, aux_bound, print_trace=True)
 
-    type_one_bound = type_one_unif_primes(int(2))
 
-    logging.info(f"Type 1 uniform bound is {type_one_bound}")
+def do_uniform_type_1(d, trial_division_bound, aux_bound):
+    mult_bound = Integer(type_one_unif_bound(d, aux_bound))
+    is_smooth, factors = is_b_smooth(mult_bound, trial_division_bound)
+    factors_str = "{" + ", ".join(str(i) for i in factors) + "}"
+    logging.info(f"A superset of the type 1 isogeny primes (trace eps = 0 and {12*d}) is: {factors_str}")
+    if not is_smooth and not is_prime(factors[-1]):
+        logging.warning(f"unable to factor by trial division, the last factor is not prime!")
 
 
-def do_uniform(d, eps):
-    tr_not_6_bd = Integer(unif_bd(d, eps))
-    logging.info(f"Trace not 6 mod 12 bound is {tr_not_6_bd}")
-    # logging.info(f"Trace not 6 mod 12 primes are {tr_not_6_bd.prime_divisors()}")
+def do_uniform_eps(d, eps, trial_division_bound, aux_bound, print_trace=False):
 
-    type_one_bound = type_one_unif_primes(d)
+    if eps == tuple(d * [0]):
+        do_uniform_type_1(d, trial_division_bound, aux_bound)
+        return
 
-    logging.info(f"Type 1 uniform bound is {type_one_bound}")
-
-
-def get_eps_from_input(tuple_as_str):
-    return tuple([Integer(t) for t in tuple_as_str.split(",")])
+    mult_bound = Integer(unif_bd(d, eps, aux_bound))
+    is_smooth, factors = is_b_smooth(mult_bound, trial_division_bound)
+    factors_str = "{" + ", ".join(str(i) for i in factors) + "}"
+    trace = sum(eps)
+    if print_trace:
+        info = f"A superset of the isogeny primes for trace eps = {trace} and {12*d - trace} is: {factors_str}"
+    else:
+        info = f"A superset of the isogeny primes for trace eps = trace {eps} = {trace} is: {factors_str}"
+    logging.info(info)
+    if not is_smooth and not is_prime(factors[-1]):
+        logging.warning(f"unable to factor by trial division, the last factor is not prime!")
```

Comment:
Well, the main reason is I wanted a function that guaranteed to succesfully determine the prime divisors of a number if all the factors are small. And that is also guaranteed to finish in a reasonable time so that our algorithm will never get stuck on factoring. What this code does is "return all prime divisors up to trial_division bound" and one integer which is what is left of the multiplicative bound after removing all the small factors (and prints a warning if it couldn't factor). This way we will never have that the code will get stuck on factoring. However as we see as soon as one computes the gcd of at least two factors there are often only small prime divisors left. So in the majority of the cases this code actually returns the prime divisors without any warning.

Reply:
nice